flightros/include/flightros/motion_planning/motion_planning.hpp

```diff
@@ -35,75 +35,47 @@
 #include "flightlib/objects/quadrotor.hpp"
 #include "flightlib/sensors/rgb_camera.hpp"
 
-// ros
-#include <ros/ros.h>
-
 namespace ob = ompl::base;
 namespace og = ompl::geometric;
 
 using namespace flightlib;
 
 namespace motion_planning {
 
-struct float2 {
-  float x, y;
-};
 struct float3 {
   float x, y, z;
 };
-struct double3 {
-  double x, y, z;
-};
-struct uint3 {
-  uint32_t x, y, z;
-};
-struct uint4 {
-  uint32_t x, y, z, w;
-};
-std::vector<float3> verts;
-float range = 1;
-bool solution_found = false;
-bool trajectory_found = false;
-
-std::vector<float3> readPointCloud();
-float3 min_bounds;
-float3 max_bounds;
-
-Eigen::Vector3d stateToEigen(const ompl::base::State *state);
-
-std::vector<ompl::base::State *> path_;
-std::vector<Eigen::Vector3d> vecs_;
-
-void getBounds();
 
-void plan();
-
-bool isStateValid(const ob::State *state);
-
-bool isInRange(float x, float y, float z);
-
-open3d::geometry::KDTreeFlann kd_tree_;
-Eigen::MatrixXd points_;
-bool searchRadius(const Eigen::Vector3d &query_point, const double radius);
-
-void executePath();
-
-// void setupQuad();
-bool setUnity(const bool render);
-bool connectUnity(void);
-
-// unity quadrotor
-std::shared_ptr<Quadrotor> quad_ptr_;
-std::shared_ptr<RGBCamera> rgb_camera_;
-QuadState quad_state_;
-
-// Flightmare(Unity3D)
-std::shared_ptr<UnityBridge> unity_bridge_ptr_;
-SceneID scene_id_{UnityScene::NATUREFOREST};
-bool unity_ready_{false};
-bool unity_render_{true};
-RenderMessage_t unity_output_;
-uint16_t receive_id_{0};
+struct Bounds {
+  float3 min;
+  float3 max;
+};
 
+class MotionPlanner {
+ public:
+  MotionPlanner();
+  ~MotionPlanner();
+  void run();
+  void readPointCloud();
+  void getBounds();
+  bool plan();
+  void executePath();
+
+ private:
+  std::vector<ompl::base::State *> path_;
+  std::vector<Eigen::Vector3d> vecs_;
```

Comment:
be consistent.  use `Vector` defined in the types.

Reply:
When using Vector<3>, the motion planning example does not compile and I did not manage to fix it yet. I encounter the following error:

Errors     << flightros:make /home/selim/Documents/catkin_ws/logs/flightros/build.make.005.log                                                                                                                    
/home/selim/Documents/catkin_ws/src/flightmare/flightros/src/motion_planning/motion_planning.cpp: In member function ‘bool motion_planning::MotionPlanner::searchRadius(flightlib::Vector<3>&, double)’:
/home/selim/Documents/catkin_ws/src/flightmare/flightros/src/motion_planning/motion_planning.cpp:265:29: error: ‘points_’ was not declared in this scope
     Vector<3> close_point = points_.col(close_point_idx);
                             ^~~~~~~
/home/selim/Documents/catkin_ws/src/flightmare/flightros/src/motion_planning/motion_planning.cpp:265:29: note: suggested alternative: ‘bounds_’
     Vector<3> close_point = points_.col(close_point_idx);
                             ^~~~~~~
                             bounds_
In file included from /usr/include/eigen3/Eigen/Core:347:0,
                 from /usr/include/eigen3/Eigen/Dense:1,
                 from /home/selim/Documents/catkin_ws/src/flightmare/flightros/include/flightros/motion_planning/motion_planning.hpp:3,
                 from /home/selim/Documents/catkin_ws/src/flightmare/flightros/src/motion_planning/motion_planning.cpp:1:
/usr/include/eigen3/Eigen/src/Core/PlainObjectBase.h: In instantiation of ‘static void Eigen::internal::conservative_resize_like_impl<Derived, OtherDerived, IsVector>::run(Eigen::DenseBase<Derived>&, Eigen::Index, Eigen::Index) [with Derived = Eigen::Matrix<float, 3, 1>; OtherDerived = Eigen::Matrix<float, 3, 1>; bool IsVector = false; Eigen::Index = long int]’:
/usr/include/eigen3/Eigen/src/Core/PlainObjectBase.h:392:60:   required from ‘void Eigen::PlainObjectBase<Derived>::conservativeResize(Eigen::Index, Eigen::Index) [with Derived = Eigen::Matrix<float, 3, 1>; Eigen::Index = long int]’
/home/selim/Documents/catkin_ws/src/flightmare/flightros/src/motion_planning/motion_planning.cpp:104:70:   required from here
/usr/include/eigen3/Eigen/src/Core/util/StaticAssert.h:32:40: error: static assertion failed: YOU_CALLED_A_DYNAMIC_SIZE_METHOD_ON_A_FIXED_SIZE_MATRIX_OR_VECTOR
     #define EIGEN_STATIC_ASSERT(X,MSG) static_assert(X,#MSG);
                                        ^
/usr/include/eigen3/Eigen/src/Core/util/StaticAssert.h:148:3: note: in expansion of macro ‘EIGEN_STATIC_ASSERT’
   EIGEN_STATIC_ASSERT(TYPE::SizeAtCompileTime==Eigen::Dynamic, \
   ^~~~~~~~~~~~~~~~~~~
/usr/include/eigen3/Eigen/src/Core/PlainObjectBase.h:923:5: note: in expansion of macro ‘EIGEN_STATIC_ASSERT_DYNAMIC_SIZE’
     EIGEN_STATIC_ASSERT_DYNAMIC_SIZE(Derived)
     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/usr/include/eigen3/Eigen/src/Core/AssignEvaluator.h: In instantiation of ‘void Eigen::internal::call_assignment_no_alias(Dst&, const Src&, const Func&) [with Dst = Eigen::Matrix<double, -1, -1>; Src = Eigen::Matrix<float, 3, 1>; Func = Eigen::internal::assign_op<double, float>]’:
/usr/include/eigen3/Eigen/src/Core/PlainObjectBase.h:728:41:   required from ‘Derived& Eigen::PlainObjectBase<Derived>::_set_noalias(const Eigen::DenseBase<OtherDerived>&) [with OtherDerived = Eigen::Matrix<float, 3, 1>; Derived = Eigen::Matrix<double, -1, -1>]’
/usr/include/eigen3/Eigen/src/Core/PlainObjectBase.h:537:19:   required from ‘Eigen::PlainObjectBase<Derived>::PlainObjectBase(const Eigen::DenseBase<OtherDerived>&) [with OtherDerived = Eigen::Matrix<float, 3, 1>; Derived = Eigen::Matrix<double, -1, -1>]’
/usr/include/eigen3/Eigen/src/Core/Matrix.h:379:29:   required from ‘Eigen::Matrix<_Scalar, _Rows, _Cols, _Options, _MaxRows, _MaxCols>::Matrix(const Eigen::EigenBase<OtherDerived>&) [with OtherDerived = Eigen::Matrix<float, 3, 1>; _Scalar = double; int _Rows = -1; int _Cols = -1; int _Options = 0; int _MaxRows = -1; int _MaxCols = -1]’
/home/selim/Documents/catkin_ws/src/flightmare/flightros/src/motion_planning/motion_planning.cpp:111:35:   required from here
/usr/include/eigen3/Eigen/src/Core/util/StaticAssert.h:32:40: error: static assertion failed: YOU_MIXED_DIFFERENT_NUMERIC_TYPES__YOU_NEED_TO_USE_THE_CAST_METHOD_OF_MATRIXBASE_TO_CAST_NUMERIC_TYPES_EXPLICITLY
     #define EIGEN_STATIC_ASSERT(X,MSG) static_assert(X,#MSG);
                                        ^
/usr/include/eigen3/Eigen/src/Core/util/XprHelper.h:816:3: note: in expansion of macro ‘EIGEN_STATIC_ASSERT’
   EIGEN_STATIC_ASSERT((Eigen::internal::has_ReturnType<ScalarBinaryOpTraits<LHS, RHS,BINOP> >::value), \
   ^~~~~~~~~~~~~~~~~~~
/usr/include/eigen3/Eigen/src/Core/AssignEvaluator.h:834:3: note: in expansion of macro ‘EIGEN_CHECK_BINARY_COMPATIBILIY’
   EIGEN_CHECK_BINARY_COMPATIBILIY(Func,typename ActualDstTypeCleaned::Scalar,typename Src::Scalar);
   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
make[2]: *** [CMakeFiles/motion_planning.dir/src/motion_planning/motion_planning.cpp.o] Error 1
make[1]: *** [CMakeFiles/motion_planning.dir/all] Error 2
make: *** [all] Error 2

Reply:
When changing them to Vector<3> I have troubles compiling it. The following error is encountered:

Errors     << flightros:make /home/selim/Documents/catkin_ws/logs/flightros/build.make.009.log                                                                                                                    
In file included from /usr/include/eigen3/Eigen/Core:347:0,
                 from /usr/include/eigen3/Eigen/Dense:1,
                 from /home/selim/Documents/catkin_ws/src/flightmare/flightros/include/flightros/motion_planning/motion_planning.hpp:3,
                 from /home/selim/Documents/catkin_ws/src/flightmare/flightros/src/motion_planning/motion_planning.cpp:1:
/usr/include/eigen3/Eigen/src/Core/AssignEvaluator.h: In instantiation of ‘void Eigen::internal::call_assignment_no_alias(Dst&, const Src&, const Func&) [with Dst = Eigen::Matrix<double, -1, -1>; Src = Eigen::Matrix<float, -1, -1>; Func = Eigen::internal::assign_op<double, float>]’:
/usr/include/eigen3/Eigen/src/Core/PlainObjectBase.h:728:41:   required from ‘Derived& Eigen::PlainObjectBase<Derived>::_set_noalias(const Eigen::DenseBase<OtherDerived>&) [with OtherDerived = Eigen::Matrix<float, -1, -1>; Derived = Eigen::Matrix<double, -1, -1>]’
/usr/include/eigen3/Eigen/src/Core/PlainObjectBase.h:537:19:   required from ‘Eigen::PlainObjectBase<Derived>::PlainObjectBase(const Eigen::DenseBase<OtherDerived>&) [with OtherDerived = Eigen::Matrix<float, -1, -1>; Derived = Eigen::Matrix<double, -1, -1>]’
/usr/include/eigen3/Eigen/src/Core/Matrix.h:379:29:   required from ‘Eigen::Matrix<_Scalar, _Rows, _Cols, _Options, _MaxRows, _MaxCols>::Matrix(const Eigen::EigenBase<OtherDerived>&) [with OtherDerived = Eigen::Matrix<float, -1, -1>; _Scalar = double; int _Rows = -1; int _Cols = -1; int _Options = 0; int _MaxRows = -1; int _MaxCols = -1]’
/home/selim/Documents/catkin_ws/src/flightmare/flightros/src/motion_planning/motion_planning.cpp:110:35:   required from here
/usr/include/eigen3/Eigen/src/Core/util/StaticAssert.h:32:40: error: static assertion failed: YOU_MIXED_DIFFERENT_NUMERIC_TYPES__YOU_NEED_TO_USE_THE_CAST_METHOD_OF_MATRIXBASE_TO_CAST_NUMERIC_TYPES_EXPLICITLY
     #define EIGEN_STATIC_ASSERT(X,MSG) static_assert(X,#MSG);
                                        ^
/usr/include/eigen3/Eigen/src/Core/util/XprHelper.h:816:3: note: in expansion of macro ‘EIGEN_STATIC_ASSERT’
   EIGEN_STATIC_ASSERT((Eigen::internal::has_ReturnType<ScalarBinaryOpTraits<LHS, RHS,BINOP> >::value), \
   ^~~~~~~~~~~~~~~~~~~
/usr/include/eigen3/Eigen/src/Core/AssignEvaluator.h:834:3: note: in expansion of macro ‘EIGEN_CHECK_BINARY_COMPATIBILIY’
   EIGEN_CHECK_BINARY_COMPATIBILIY(Func,typename ActualDstTypeCleaned::Scalar,typename Src::Scalar);
   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
make[2]: *** [CMakeFiles/motion_planning.dir/src/motion_planning/motion_planning.cpp.o] Error 1
make[1]: *** [CMakeFiles/motion_planning.dir/all] Error 2
make: *** [all] Error 2

Reply:
Any idea on how to solve this? It seems there are some mixed numeric types.